src/sinks/aws_cloudwatch_logs/mod.rs

```diff
@@ -235,36 +243,27 @@ impl Service<PartitionInnerBuffer<Vec<InputLogEvent>, CloudwatchKey>>
         let svc = if let Some(svc) = &mut self.clients.get_mut(&key) {
             svc.clone()
         } else {
-            let svc = {
-                let policy = self.request_settings.retry_policy(CloudwatchRetryLogic);
-
-                let cloudwatch = CloudwatchLogsSvc::new(&self.config, &key, self.client.clone());
-                let timeout = Timeout::new(cloudwatch, self.request_settings.timeout);
-
-                let buffer = Buffer::new(timeout, 1);
-                let retry = Retry::new(policy, buffer);
-
-                let rate = Rate::new(
+            let svc = ServiceBuilder::new()
+                .buffer(self.request_settings.in_flight_limit.unwrap())
+                .concurrency_limit(1)
```

Comment:
This doesn't look right. `concurrency_limit()` should apply the `in_flight_limit`. Maybe this is the cause of the integration test issue you mentioned.

Also the same note about unwrapping here.

Reply:
I think that integration test issue caused by not calling `poll_complete` here: https://github.com/timberio/vector/blob/8d1995d0d6243624a540b8dfc824e51ae5d77044/src/sinks/util/sink.rs#L376-L389
`poll_send` probably get infinity loop, not sure that this is correct and not sure why this showed after autoconcurrency merge
https://github.com/timberio/vector/blob/8d1995d0d6243624a540b8dfc824e51ae5d77044/src/sinks/util/sink.rs#L517-L519

Reply:
That doesn't address the core concern, which is that this appears to be setting the buffer size to the configured concurrency limit, and fixes the concurrency to 1. Am I missing something? I think I understand the fixed concurrency due to this being a sub-service (though it then doesn't need the concurrency layer at all except for convenience of using the `Svc` type alias) but the buffer setting still looks wrong to me.

Reply:
We need to set the concurrency to 1 because request requires token from previous request (`poll_ready` depends from this):
https://github.com/timberio/vector/blob/8b719cd6bc47ded6f0be46e46ca3eb06725d9b5a/src/sinks/aws_cloudwatch_logs/mod.rs#L353-L369
In the same time, I changed buffer to `in_flight_limit` before concurrency because `poll_ready` on service always return `Poll::Ready` and inner service should have `in_flight_limit` capacity.
https://github.com/timberio/vector/blob/8b719cd6bc47ded6f0be46e46ca3eb06725d9b5a/src/sinks/aws_cloudwatch_logs/mod.rs#L233-L235

Reply:
Ok, I that makes more sense. This might be worth a bit of internal documentation here.

Reply:
Added!